CDAP-8107 File metadata manager refactoring and log handler changes

gsps1 opened the pull request:
Refactored write meta logic into FileMetadataWriter and read meta data logic to FileMetadataReader. 
so the writes can follow the new format, while reads can support reading logs in old and new format. 

Changes in LogFileHandler to use FileMetadataReader and LogLocation classes to read log. 

Test cases and minor changes.

JIRA : https://issues.cask.co/browse/CDAP-8107

gsps1 committed:
CDAP-8107 File metadata manager refactoring with reader/writer and log handler changes

CuriousVini commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/write/FileMetaDataManager.java:
```diff
@@ -109,31 +105,6 @@ public void writeMetaData(final LoggingContext loggingContext,
-      @Override
-      public void apply(Table table) throws Exception {
-        // add column version prefix for new format
-        byte[] columnKey = Bytes.add(COLUMN_PREFIX_VERSION, Bytes.toBytes(eventTimeMs), Bytes.toBytes(currentTimeMs));
```
Unused: `COLUMN_PREFIX_VERSION` on line 68/64

CuriousVini commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/write/FileMetaDataManager.java:
```diff
@@ -109,31 +105,6 @@ public void writeMetaData(final LoggingContext loggingContext,
-      public void apply(Table table) throws Exception {
-        // add column version prefix for new format
-        byte[] columnKey = Bytes.add(COLUMN_PREFIX_VERSION, Bytes.toBytes(eventTimeMs), Bytes.toBytes(currentTimeMs));
-        table.put(getRowKey(identifier), columnKey, Bytes.toBytes(location.toURI().toString()));
```
Unused method definition present for `getRowKey`

CuriousVini commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/write/FileMetaDataManager.java:
```diff
@@ -190,50 +161,6 @@ public Void call() throws Exception {
-   * @param logPathIdentifier logging context identifier.
-   * @return List of {@link LogLocation}
-   */
-  public List<LogLocation> listFiles(final LogPathIdentifier logPathIdentifier) throws Exception {
```
So we are only moving code used by new `CDAPLogAppender` remainder still remains in `FileMetadataManager`? Because we are planning to eventually remove that class?

gsps1 replied:
yes.

gsps1 commented:
build passed : http://builds.cask.co/browse/CDAP-DUT5354-3

CuriousVini commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/read/FileLogReader.java:
```diff
@@ -88,24 +73,18 @@ public void getLogNext(final LoggingContext loggingContext, final ReadRange read
       }
 
-      List<Location> filesInRange = getFilesInRange(sortedFiles, readRange.getFromMillis(), readRange.getToMillis());
-      AvroFileReader logReader = new AvroFileReader(schema);
```
Why dont we use AvroFileReader rather than copy pasting the same code in `LogLocation` ?

gsps1 replied:
we will get rid of AvroFileReader and just use LogLocation.

CuriousVini commented on cdap-watchdog/src/test/java/co/cask/cdap/logging/save/LogSaverPluginTest.java:
```diff
@@ -177,10 +172,15 @@ public void testPlugin() throws Exception {
+                               injector.getInstance(Impersonator.class),
+                               "log.meta");
+
+      waitTillLogSaverDone(fileMetadataReader, loggingContext, "Test log message 59 arg1 arg2");
```
we should make changes to reader of LogSaverTest#waitTillLogSaverDone too

gsps1 replied:
good point, i made changes to that and also `LogSaverPluginTest` and removed the listFiles method in `FileMetadataManager`. so we only use `FileMetadataReader` for listing files.

gsps1 commented:
Addressed comments, changed branch to release/4.1 and resolved conflicts. I rebased commits so it would be easier to do conflict resolution. @CuriousVini could you please review again. thanks

CuriousVini commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/read/FileMetadataReader.java:
```diff
@@ -0,0 +1,225 @@
+          }
+        }
+
+        return getFilesInRange(files, startTimestampMs);
```
I think it would be better to do `getFilesInRange` outside the transaction similar to `Processor` in `FileMetadataManager#scanFiles()`

gsps1 replied:
good point, changed.

CuriousVini commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/read/FileMetadataReader.java:
```diff
@@ -0,0 +1,225 @@
+
+        for (Map.Entry<byte[], byte[]> entry : cols.getColumns().entrySet()) {
+          // the location can be any location from on the filesystem for custom mapped namespaces
+          if (entry.getKey().length == 8) {
```
It would be good to add more comments around `entry.getKey().length` length being 8 vs 17.

gsps1 replied:
added some comments

CuriousVini commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/write/LogLocation.java:
```diff
@@ -111,6 +118,139 @@ public long getFileCreationTimeMs() {
   }
 
+  /**
+   * @param logFilter filter
```
description?

CuriousVini commented on cdap-watchdog/src/test/java/co/cask/cdap/logging/save/LogSaverPluginTest.java:
```diff
@@ -489,21 +492,21 @@ public void run() {
+      LogLocation lastEntry = files.get(files.size() - 1);
       if (lastEntry != null) {
-        Location latestFile = lastEntry.getValue();
-        AvroFileReader logReader = new AvroFileReader(LogSchema.LoggingEvent.SCHEMA);
```
Should we delete `AvroFileReader` since we do not use it anywhere?

gsps1 replied:
good point, deleted.

CuriousVini commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/read/FileMetadataReader.java:
```diff
@@ -0,0 +1,225 @@
+          smallestTimestamp = eventTimestamp;
+        }
+      }
+      filteredList.add(0, logLocation);
```
Why do we always add `logLocation` to index 0?

gsps1 replied:
we process from end of list after sorting, there are comments above

CuriousVini commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/read/FileMetadataReader.java:
```diff
@@ -0,0 +1,225 @@
+            }
+          } else if (entry.getKey().length == 17) {
+            long eventTimestamp = Bytes.toLong(entry.getKey(), 1, Bytes.SIZEOF_LONG);
+            if (!(eventTimestamp > endTimestampMs)) {
```
Why can't we filter based on `startTimestampMs` and `endTimestampMs` here? why do we have a separate method to do filtering only on `startTimestampMs`

gsps1 replied:
the reason is the "timestamp" we use in column key in the event timestamp of the first event written to a file, so if end timestamp in query is smaller than this event timestamp, we can be sure that this file doesn't contain logs in our query range, however the same cannot be true for start timestamp filtering.

gsps1 replied:
file-name| event-timestamp |creation-timestamp   
---------|-----------------|----------------------
f0|100|t0
f1|100|t1
f2|200|t2
f3|500|t3

If my query range is 180 - 300.
then i can be sure that the last file `f3` doesn't have logs in my time range as the starting event's timestamp is larger than my query's end time stamp.

`f2` is definitely in range.
 `f1` may have logs from time range 180-200
`f0` out of range  - `f1` start time stamp is 100 but created after `f0`. so all the logs in `f0` would have timestamp 100. so out of range.

this is why we have extra processing for filtering by start timestamp. we sort and process in reverse order.

CuriousVini commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/read/FileMetadataReader.java:
```diff
@@ -0,0 +1,225 @@
+      throw Transactions.propagate(e);
+    }
+  }
+
```
nit: extra lines

CuriousVini reviewed (commented):
Minor comments and questions around sorting and filtering logic

gsps1 commented:
@CuriousVini addressed comments.

gsps1 commented:
build running : http://builds.cask.co/browse/CDAP-RUT485-1

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/write/FileMetaDataWriter.java:
```diff
@@ -0,0 +1,118 @@
+    LOG.debug("Writing meta data for logging context {} with eventTime {} currentTime {} and location {}",
+              identifier.getRowKey(), eventTimeMs, currentTimeMs, location);
+
+    txExecute(transactional, new TxRunnable() {
```
This method already has `throws Exception` in the signature, why need to wrap it with `txExecute` (which cause and rethrow with runtime exception)? I saw similar thing in the reader class as well.

chtyim replied:
If you want to unwrap the transaction exception appropriately, you should have the `txExecute` method has `throws Exception` and then in the `catch` block, do a `throw Transactions.propagate(e, Exception.class);`

gsps1 commented:
will update the PR after this https://github.com/caskdata/cdap/pull/7847 is merged.

gsps1 commented:
@CuriousVini @chtyim updated the PR after rebase. please review.

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/FileMetaDataReader.java:
```diff
@@ -0,0 +1,180 @@
+  private final Impersonator impersonator;
+
+  @Inject
+  public FileMetaDataReader(final DatasetFramework datasetFramework,
```
For `@Inject` constructor, better to make it non-public (just leave it as package private).

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/FileMetaDataReader.java:
```diff
@@ -0,0 +1,180 @@
+            final Row cols = table.get(getRowKey(logPathIdentifier));
+
+            if (cols.isEmpty()) {
+              //noinspection unchecked
```
Why need this?? Also better to return `Collections.emptyList()`.

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/FileMetaDataReader.java:
```diff
@@ -0,0 +1,180 @@
+              return new ArrayList<>();
+            }
+
+            final List<LogLocation> files = new ArrayList<>();
```
You don't need `final`, right?

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/FileMetaDataReader.java:
```diff
@@ -0,0 +1,180 @@
+              // new row-key format length is 17 bytes (version_byte, event timestamp (8b) and current timestamp (8b)
+              if (entry.getKey().length == 8) {
+                long eventTimestamp = Bytes.toLong(entry.getKey());
+                if (!(eventTimestamp > endTimestampMs)) {
```
maybe easier to read if you have `if (eventTimestamp <= endTimestampMs)`

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/FileMetaDataReader.java:
```diff
@@ -0,0 +1,180 @@
+                }
+              } else if (entry.getKey().length == 17) {
+                long eventTimestamp = Bytes.toLong(entry.getKey(), 1, Bytes.SIZEOF_LONG);
+                if (!(eventTimestamp > endTimestampMs)) {
```
Same as above.

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/FileMetaDataReader.java:
```diff
@@ -0,0 +1,180 @@
+                                            Bytes.toLong(entry.getKey(), 9, Bytes.SIZEOF_LONG),
+                                            rootLocationFactory.create(new URI(Bytes.toString(entry.getValue()))),
+                                            logPathIdentifier.getNamespaceId(), impersonator));
+                }
```
How about other key length? Shall we log a warn?

gsps1 replied:
added warn message

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/FileMetaDataReader.java:
```diff
@@ -0,0 +1,180 @@
+                if (!(eventTimestamp > endTimestampMs)) {
+                  // new format
+                  files.add(new LogLocation(LogLocation.VERSION_1,
+                                            // skip the first (version) byte
```
This comment is out of place. I believe you want to put it above line 127.

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/FileMetaDataReader.java:
```diff
@@ -0,0 +1,180 @@
+      return files;
+    }
+    // sort the list
+    Collections.sort(files, LOG_LOCATION_COMPARATOR);
```
Isn't it sorted already when reading from the backing store? The columns are of fixed length and is having (exclude the constant version byte) `eventTime` as the first 8 bytes, followed by the `createTime`

gsps1 replied:
we do `getColumns` for the rowkey, which is a map.

chtyim replied:
It actually is a sorted map... The API is not defined well... Anyway, keep this sort for now.

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/FileMetaDataReader.java:
```diff
@@ -0,0 +1,180 @@
+    // we will return the file with most recent creation time - which is the expected behavior.
+    // if we reach the beginning of the list, we return the entire list.
+    List<LogLocation> filteredList = new ArrayList<>();
+    for (int i = files.size() - 1; i >= 0; i--) {
```
You can do a `for (LogLocation logLocation : Lists.reverse(files)) {`

gsps1 commented:
@chtyim addressed comments.

chtyim commented:
LGTM. Please squash and merge.

gsps1 commented:
waiting for build : http://builds.cask.co/browse/CDAP-RUT485-5

gsps1 commented:
merging it. build failing with unrelated error, but have had successful build of this branch earlier. also since the new log framework PR is merged, we need this for retrieving logs in new format.
